test/extensions/transport_sockets/tls/rsa_private_key_method_provider.cc

```diff
@@ -0,0 +1,296 @@
+RsaPrivateKeyMethodProvider::getBoringSslPrivateKeyMethod() {
+  return method_;
+}
+
+RsaPrivateKeyConnection::RsaPrivateKeyConnection(SSL* ssl, Ssl::PrivateKeyConnectionCallbacks& cb,
+                                                 Event::Dispatcher& dispatcher,
+                                                 bssl::UniquePtr<EVP_PKEY> pkey,
+                                                 RsaPrivateKeyConnectionTestOptions& test_options)
+    : test_options_(test_options), cb_(cb), dispatcher_(dispatcher), pkey_(move(pkey)) {
+  SSL_set_ex_data(ssl, RsaPrivateKeyMethodProvider::ssl_rsa_connection_index, this);
+}
+
+Ssl::PrivateKeyConnectionPtr RsaPrivateKeyMethodProvider::getPrivateKeyConnection(
+    SSL* ssl, Ssl::PrivateKeyConnectionCallbacks& cb, Event::Dispatcher& dispatcher) {
+  bssl::UniquePtr<BIO> bio(
+      BIO_new_mem_buf(const_cast<char*>(private_key_.data()), private_key_.size()));
+  bssl::UniquePtr<EVP_PKEY> pkey(PEM_read_bio_PrivateKey(bio.get(), nullptr, nullptr, nullptr));
+  if (pkey == nullptr) {
+    return nullptr;
+  }
+
+  return std::make_unique<RsaPrivateKeyConnection>(ssl, cb, dispatcher, move(pkey), test_options_);
+}
+
+RsaPrivateKeyMethodProvider::RsaPrivateKeyMethodProvider(
+    const ProtobufWkt::Struct& config,
+    Server::Configuration::TransportSocketFactoryContext& factory_context) {
+
+  std::string private_key_path;
+
+  if (RsaPrivateKeyMethodProvider::ssl_rsa_connection_index == -1) {
+    RsaPrivateKeyMethodProvider::ssl_rsa_connection_index =
+        SSL_get_ex_new_index(0, nullptr, nullptr, nullptr, nullptr);
+  }
+
+  for (auto& value_it : config.fields()) {
+    auto& value = value_it.second;
+    if (value_it.first == "private_key_file" &&
+        value.kind_case() == ProtobufWkt::Value::kStringValue) {
+      private_key_path = value.string_value();
+    }
+    if (value_it.first == "sync_mode" && value.kind_case() == ProtobufWkt::Value::kBoolValue) {
+      test_options_.sync_mode_ = value.bool_value();
+    }
+    if (value_it.first == "crypto_error" && value.kind_case() == ProtobufWkt::Value::kBoolValue) {
+      test_options_.crypto_error_ = value.bool_value();
+    }
+    if (value_it.first == "method_error" && value.kind_case() == ProtobufWkt::Value::kBoolValue) {
+      test_options_.method_error_ = value.bool_value();
+    }
+    if (value_it.first == "async_method_error" &&
+        value.kind_case() == ProtobufWkt::Value::kBoolValue) {
+      test_options_.async_method_error_ = value.bool_value();
+    }
+    if (value_it.first == "expected_operation" &&
+        value.kind_case() == ProtobufWkt::Value::kStringValue) {
+      if (value.string_value() == "decrypt") {
+        test_options_.decrypt_expected_ = true;
+      } else if (value.string_value() == "sign") {
+        test_options_.sign_expected_ = true;
+      }
+    }
+  }
+
+  private_key_ = factory_context.api().fileSystem().fileReadToEnd(private_key_path);
+
+  method_ = std::make_shared<SSL_PRIVATE_KEY_METHOD>();
```

Comment:
Is there any way to reuse the BoringSSL test code directly? I.e. via a `#include`?

Reply:
I'll investigate this. The problem is that we want to inject errors to the methods to test the various failure modes.

Reply:
It might be possible to make changes to BoringSSL and then bump our dependency version. The maintainers are Googlers, so we can try and coordinate on this if necessary.